### PR TITLE
Expose `kube-apiservers` of `Shoots` with trusted wildcard certificate

### LINKED ISSUE #7233: Use wildcard certificate to expose `kube-apiserver`s of `Shoot` with trusted certificate
**How to categorize this issue?**

/area networking usability
/kind enhancement

**What would you like to be added**:
Operators can already provide a wildcard certificate for the seed cluster's ingress domain which is used to expose the observability components with a trusted certificate. This functionality is undocumented.

The `gardenlet`'s `Shoot` controller copies the wildcard certificate secret into the shoot namespaces: https://github.com/gardener/gardener/blob/759c2641dab95cba6653924e3462aeec2e1fe56d/pkg/operation/botanist/secrets.go#L309-L335
When constructing the `Ingress` resources for the observability components, this secret is used in `spec.tls[].secretName`: https://github.com/gardener/gardener/blob/759c2641dab95cba6653924e3462aeec2e1fe56d/pkg/operation/botanist/monitoring.go#L164-L180
https://github.com/gardener/gardener/blob/759c2641dab95cba6653924e3462aeec2e1fe56d/charts/seed-monitoring/charts/alertmanager/templates/ingress.yaml#L17-L21

Let's enhance `gardenlet` to also create an `Ingress` resource for the `kube-apiserver` of shoots with a subdomain of the seed's ingress domain (similar like it's done for the observability components). Use this wildcard certificate secret to expose it. In order for this to work, `kube-apiserver` must be additionally configured with `--tls-sni-cert-key`.
From https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/:

> A pair of x509 certificate and private key file paths, optionally suffixed with a list of domain patterns which are fully qualified domain names, possibly with prefixed wildcard segments. The domain patterns also allow IP addresses, but IPs should only be used if the apiserver has visibility to the IP address requested by a client. If no domain patterns are provided, the names of the certificate are extracted. Non-wildcard matches trump over wildcard matches, explicit domain patterns trump over extracted names. For multiple key/certificate pairs, use the --tls-sni-cert-key multiple times. Examples: "example.crt,example.key" or "foo.crt,foo.key:*.foo.com,foo.com".

**Why is this needed**:
The https://github.com/gardener/terminal-controller-manager could benefit from it since it currently sets this up manually.

### PR DESCRIPTION
**Pull request by oliver-goetz**
**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:
This PR exposes the `kube-apiservers` of `Shoot` with the trusted wildcard certificate via Ingress if it exists. 
Their addresses look like: `https://api-<project-name>--<shoot-name>.<seed-ingress-domain>`.
If there is no wildcard certificate on the seed the ingress object is not created.
This endpoint **should not** be used by end-users - they should continue using the URL provided by Gardener in their kubeconfigs.

**Which issue(s) this PR fixes**:
Fixes #7233 

**Special notes for your reviewer**:

**Release note**:

```feature operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	// Host is the host where the kube-apiserver should be exposed.
+	Host string
+	// IngressClassName is the name of the ingress class.
+	IngressClassName *string
+	// ServiceName is the name of the service the ingress is using
+	ServiceName string
+}
+
+// NewIngress creates a new instance of Deployer for the ingress used to expose the kube-apiserver.
+func NewIngress(c client.Client, objectKey client.ObjectKey, values IngressValues) component.Deployer {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go]: Why do I need to pass the object key/why is this configurable?

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+				},
+			},
+			TLS: []networkingv1.IngressTLS{{Hosts: []string{i.values.Host}}},
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go]:
```suggestion
	_, err := controllerutils.GetAndCreateOrMergePatch(ctx, i.client, ingress, func() error {
		metav1.SetMetaDataAnnotation(&ingress.ObjectMeta, nginxIngressSSLPassthrough, "true")
		ingress.Labels = utils.MergeStringMaps(ingress.Labels, getLabels())
		pathType := networkingv1.PathTypePrefix
		ingress.Spec = networkingv1.IngressSpec{
			IngressClassName: i.values.IngressClassName,
			Rules: []networkingv1.IngressRule{
				{
					Host: i.values.Host,
					IngressRuleValue: networkingv1.IngressRuleValue{
						HTTP: &networkingv1.HTTPIngressRuleValue{
							Paths: []networkingv1.HTTPIngressPath{
								{
									Backend: networkingv1.IngressBackend{
										Service: &networkingv1.IngressServiceBackend{
											Name: i.values.ServiceName,
											Port: networkingv1.ServiceBackendPort{
												Number: kubeapiserverconstants.Port,
											},
										},
									},
									Path:     "/",
									PathType: &pathType,
								},
							},
						},
					},
				},
			},
			TLS: []networkingv1.IngressTLS{{Hosts: []string{i.values.Host}}},
		}
		return nil
	})

	return err
```

File: pkg/operation/botanist/kubeapiserverexposure.go [hunk 3]
```diff
@@ -151,3 +151,32 @@ func (b *Botanist) setAPIServerServiceClusterIP(clusterIP string) {
...
+func (b *Botanist) DeployKubeAPIServerIngress(ctx context.Context) error {
+	// Do not deploy ingress if there is no wildcard certificate
+	if b.ControlPlaneWildcardCert == nil {
+		return b.DestroyKubeAPIServerIngress(ctx)
+	}
+	ingressClass, err := gardenerutils.ComputeNginxIngressClassForSeed(b.Seed.GetInfo(), b.Seed.GetInfo().Status.KubernetesVersion)
+	if err != nil {
+		return err
+	}
+	return kubeapiserverexposure.NewIngress(
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserverexposure.go:162]: Instead of creating the component on the fly in these functions, can you add it to https://github.com/gardener/gardener/blob/d21b3ccfa1b6d76e483ba83b6f292d232ca35e55/pkg/operation/shoot/types.go#L117-L132 and create it [here](https://github.com/gardener/gardener/blob/d21b3ccfa1b6d76e483ba83b6f292d232ca35e55/pkg/operation/botanist/botanist.go#L133-L135)?

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by plkokanov** [comment]
In #7233 it is mentioned that this functionality is not documented (I also noticed that there is no release note):
> Operators can already provide a wildcard certificate for the seed cluster's ingress domain which is used to expose the observability components with a trusted certificate. **This functionality is undocumented.**

Is there a reason for this? If not, could we use this PR to also add appropriate documentation and a release note?

**Comment by rfranzke** [comment]
The reason is described in https://github.com/gardener/gardener/issues/7233#issuecomment-1358934692

File: pkg/operation/common/types.go [hunk 4]
```diff
@@ -18,6 +18,9 @@ const (
 	// VPNTunnel dictates that VPN is used as a tunnel between seed and shoot networks.
 	VPNTunnel string = "vpn-shoot"
 
+	// KubeAPIServerUsersPrefix is a constant for a prefix used for the users KubeAPIServer instance using a TLS certificate from a trusted origin.
```
Comment (plkokanov) [inline pkg/operation/common/types.go]:
```suggestion
	// KubeAPIServerPrefix is a constant for a prefix used for the KubeAPIServer instance which uses a TLS certificate from a trusted origin.
```
There is no users and operators distinction for the kube-apiserver ingress If I'm not mistaken.

File: pkg/operation/common/types.go [hunk 5]
```diff
@@ -18,6 +18,9 @@ const (
...
+	KubeAPIServerUsersPrefix = "api"
```
Comment (plkokanov) [inline pkg/operation/common/types.go]:
```suggestion
	KubeAPIServerPrefix = "api"
```

File: pkg/operation/botanist/kubeapiserverexposure.go [hunk 6]
```diff
@@ -151,3 +151,32 @@ func (b *Botanist) setAPIServerServiceClusterIP(clusterIP string) {
...
+		b.SeedClientSet.Client(),
+		client.ObjectKey{Name: v1beta1constants.DeploymentNameKubeAPIServer, Namespace: b.Shoot.SeedNamespace},
+		kubeapiserverexposure.IngressValues{
+			ServiceName:      v1beta1constants.DeploymentNameKubeAPIServer,
+			Host:             b.ComputeKubeAPIServerHost(),
+			IngressClassName: &ingressClass,
+		}).Deploy(ctx)
+}
+
+// DestroyKubeAPIServerIngress deploys ingress for the kube-apiserver.
```
Comment (plkokanov) [inline pkg/operation/botanist/kubeapiserverexposure.go]:
```suggestion
// DestroyKubeAPIServerIngress destroys the ingress for the kube-apiserver.
```

File: pkg/operation/botanist/kubeapiserverexposure.go [hunk 7]
```diff
@@ -151,3 +151,32 @@ func (b *Botanist) setAPIServerServiceClusterIP(clusterIP string) {
...
+	ingressClass, err := gardenerutils.ComputeNginxIngressClassForSeed(b.Seed.GetInfo(), b.Seed.GetInfo().Status.KubernetesVersion)
+	if err != nil {
+		return err
+	}
+	return kubeapiserverexposure.NewIngress(
+		b.SeedClientSet.Client(),
+		client.ObjectKey{Name: v1beta1constants.DeploymentNameKubeAPIServer, Namespace: b.Shoot.SeedNamespace},
+		kubeapiserverexposure.IngressValues{
+			ServiceName:      v1beta1constants.DeploymentNameKubeAPIServer,
+			Host:             b.ComputeKubeAPIServerHost(),
```
Comment (plkokanov) [inline pkg/operation/botanist/kubeapiserverexposure.go:167]: When you create a component for thee kube-apiserver ingress as @rfranzke suggested you could either pass a function argument, similar to what we do for the SNI component: https://github.com/gardener/gardener/blob/47822ef2d6e0b3f2a6403aa4e54f3fa8e3c27405/pkg/operation/botanist/kubeapiserverexposure.go#L73-L88
or create an additional setter for the value returned by the `ComputeKubeAPIServerHost()` function

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	nginxIngressSSLPassthrough = "nginx.ingress.kubernetes.io/ssl-passthrough"
+)
+
+// IngressValues configure the kube-apiserver ingress.
+type IngressValues struct {
+	// Host is the host where the kube-apiserver should be exposed.
+	Host string
+	// IngressClassName is the name of the ingress class.
+	IngressClassName *string
+	// ServiceName is the name of the service the ingress is using
```
Comment (plkokanov) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go]:
```suggestion
	// ServiceName is the name of the service the ingress is using.
```

File: pkg/operation/botanist/kubeapiserverexposure.go [hunk 9]
```diff
@@ -151,3 +151,32 @@ func (b *Botanist) setAPIServerServiceClusterIP(clusterIP string) {
 		},
 	)
 }
+
+// DeployKubeAPIServerIngress deploys ingress for the kube-apiserver.
+func (b *Botanist) DeployKubeAPIServerIngress(ctx context.Context) error {
```
Comment (plkokanov) [inline pkg/operation/botanist/kubeapiserverexposure.go]: Unit tests for these functions might be useful to make sure that the `ingressClass` is specified proprely and to make sure that when `b.ControlPlaneWildcardCert` is false, the ingress is destroyed.

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/kubeapiserverexposure.go (hunk 7 above)
Comment (rfranzke) [reply pkg/operation/botanist/kubeapiserverexposure.go:167]: Is this additional complexity really needed here for the ingress?

Comment (plkokanov) [reply pkg/operation/botanist/kubeapiserverexposure.go:167]: True, after another look I guess `ComputeKubeAPIServerHost()` can simply be called when the component is created where you suggested.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go (hunk 1 above)
Comment (oliver-goetz) [reply pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_ingress.go]: Only the namespace is required, the name can be a constant.

File: pkg/operation/botanist/kubeapiserverexposure.go (hunk 9 above)
Comment (oliver-goetz) [reply pkg/operation/botanist/kubeapiserverexposure.go]: Too late for testing now. I'll check it tomorrow 😄

Comment (oliver-goetz) [reply pkg/operation/botanist/kubeapiserverexposure.go]: ok, here they are

File: pkg/operation/common/types.go [hunk 10]
```diff
@@ -18,6 +18,9 @@ const (
 	// VPNTunnel dictates that VPN is used as a tunnel between seed and shoot networks.
 	VPNTunnel string = "vpn-shoot"
 
+	// KubeAPIServerPrefix is a constant for a prefix used for the users KubeAPIServer instance which uses a TLS certificate from a trusted origin.
```
Comment (plkokanov) [inline pkg/operation/common/types.go]: Nit:
```suggestion
	// KubeAPIServerPrefix is a constant for a prefix used for the KubeAPIServer instance which uses a TLS certificate from a trusted origin.
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
Thanks for addressing the review suggestions. Just one minor nit left.

File: pkg/operation/botanist/kubeapiserverexposure.go [hunk 11]
```diff
@@ -151,3 +151,38 @@ func (b *Botanist) setAPIServerServiceClusterIP(clusterIP string) {
...
+}
+
+// DestroyKubeAPIServerIngress destroys the ingress for the kube-apiserver.
+func (b *Botanist) DestroyKubeAPIServerIngress(ctx context.Context) error {
+	return kubeapiserverexposure.NewIngress(
+		b.SeedClientSet.Client(),
+		b.Shoot.SeedNamespace,
+		kubeapiserverexposure.IngressValues{},
+	).Destroy(ctx)
+}
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserverexposure.go]:
```suggestion
	if b.ControlPlaneWildcardCert == nil {
		return b.Shoot.Components.ControlPlane.KubeAPIServerIngress.Destroy(ctx)
	}
	return b.Shoot.Components.ControlPlane.KubeAPIServerIngress.Deploy(ctx)
}
```

File: pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go [hunk 12]
```diff
@@ -533,6 +533,11 @@ func (r *Reconciler) runDeleteShootFlow(ctx context.Context, o *operation.Operat
 			Fn:           flow.TaskFn(botanist.Shoot.Components.ControlPlane.KubeAPIServerSNI.Destroy).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(waitUntilKubeAPIServerDeleted),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying Kubernetes API server ingress with trusted certificate",
+			Fn:           flow.TaskFn(botanist.DestroyKubeAPIServerIngress),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go]:
```suggestion
			Fn:           botanist.Shoot.Components.ControlPlane.KubeAPIServerIngress.Destroy,
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 13]
```diff
@@ -546,6 +546,11 @@ func (b *Botanist) computeKubeAPIServerSNIConfig() kubeapiserver.SNIConfig {
 		}
 	}
 
+	// Add control plane wildcard certificate to TLS SNI config if it is available.
+	if b.ControlPlaneWildcardCert != nil {
+		config.TLS = append(config.TLS, kubeapiserver.TLSSNIConfig{SecretName: &b.ControlPlaneWildcardCert.Name, DomainPatterns: []string{b.ComputeKubeAPIServerHost()}})
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go:552]: Does it make sense to add a small unit test that ensures that the `TLS` field is properly set when the `ControlPlaneWildcardCert` is not `nil`?

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
